### PR TITLE
Now the CSE tests use the pipeline-based approach

### PR DESCRIPTION
**Pull request by Anabra**

### CODE REVIEW

**Comment by csabahruska** [comment]
Why is this good? It adds more dependencies to test machinery.
Why is this required?

**Comment by Anabra** [comment]
Not required. Just thought that it would be better to use a standardized approach when testing. However, it brings yet another dependency for this module, so it might not be a good idea if we want to test each part of the compiler separately.

File: grin/test/Transformations/Optimising/CSESpec.hs [hunk 1]
```diff
@@ -15,10 +18,16 @@ import Transformations.EffectMap
 runTests :: IO ()
 runTests = hspec spec
 
-cseOptNoEff (tyEnv, exp) = commonSubExpressionElimination tyEnv mempty exp
+pipelineSrcWithCtx :: ((TypeEnv, Exp) -> (TypeEnv, Exp)) ->
+                      TypeEnv -> Exp -> Exp ->
+                      [PipelineStep] -> Pipeline
+pipelineSrcWithCtx ctx tyEnv before after =
+  pipelineSrc (snd $ ctx (tyEnv, before)) (snd $ ctx (tyEnv, after))
```
Comment (andorp) [inline grin/test/Transformations/Optimising/CSESpec.hs]: I think this extra indirection makes the readability worse. Also I am not sure anymore that the ctx idea worthwhile.

But definetly we need to run a cleanup round around the first release, until than I would postpone any non-trivial cleanup.

**Comment by andorp** [comment]
@Anabra: What is your final say? Do you want to merge this PR or reject it? I have a soft objection about the extra indirection, but I don't mind either having it in the module...

**Comment by Anabra** [comment]
The implementation in this PR only makes sense if the pipeline consists of several different stages. Here, it is only CSE. Furthermore, it introduces yet another dependency for the module. In hindsight, this is not a welcome change.

As for the context-based testing approach, I really like the idea. Write one test, get several. The code is a bit difficult to understand, but reading it through once or twice is enough to know how to use it. I think it should stay.

**Pull request closed without merging**